storage_sync2: storage_sync: handle task_mgr::is_shutdown_requested()

problame opened the pull request:
storage_sync: handle task_mgr::is_shutdown_requested()                      
                                                                            
- Introduce another UploadQueue::ShutDown enum variant to indicate the state
  where the UploadQueue is shut down.                                       
- In perform_upload_task, check for task_mgr::is_shutdown_requested()       
  before going into expoential backoff.                                     
  If we are requested to shut down, the first in-progress tasks that        
  notices the shutdown request transitions the queue from                   
  UploadQueue::Initialized to UploadQueue::ShutDown.                        
  This involves dropping all the queued ops that are not yet                
  in progress, which conveniently unblocks wait_completion() calls          
  that are waiting for their barrier to be executed.                        
  They will receive an Err(), and do something sensible.                    
                                                                            
Right now, wait_completion() is only used by tests, but I                   
suspect that we should be using it in wherever we                           
delete layer files, e.g., GC and compaction, as explained                   
in the storage_sync.rs block comment section "Consistency".                 
                                                                            
This change als fixes                                                       
test_timeline_deletion_with_files_stuck_in_upload_queue                     
which I added in the previous commit.                                       
Before, timeline delete would wait until all in-progress                    
tasks and queued tasks were done.                                           
If, like in the test, a task was stuck due to upload error,                 
timeline deletion would wait forever                                        
Hence, handling task_mgr::is_shutdown_requested() is sufficient             
to resolve this particular problem.

problame commented on pageserver/src/storage_sync.rs:
```diff
@@ -800,72 +823,93 @@ impl RemoteTimelineClient {
+                // FIXME this captures the parent span, i.e. log messages inside the task will look like so:
+                // 2022-11-24T17:34:35.707155Z  INFO layer flush task{tenant=81892ec7b59f9ac3f06f5ab5d8c1fe3d timeline=82f9c121ee5778016485e775550e8c61}:flush_frozen_layer{tenant_id=81892ec7b59f9ac3f06f5ab5d8c1fe3d timeline_id=82f9c121ee5778016485e775550e8c61 layer=inmem-0000000001696629-00000000016AFA39}:remote_upload{tenant=81892ec7b59f9ac3f06f5ab5d8c1fe3d timeline=82f9c121ee5778016485e775550e8 ...
+                // How to avoid this?
+                .instrument(info_span!("remote_upload", tenant = %self.tenant_id, timeline = %self.timeline_id, upload_task_id = %task_id)),
```
@koivunej  I bet you know what to do here?

hlinnaka replied:
You can set the parent span in the `info_span!` macro like this:
```
.instrument(info_span!(parent: None, "remote_upload", tenant = %self.tenant_id, timeline = %self.timeline_id, upload_task_id = %task_id)),
```

hlinnaka commented on pageserver/src/storage_sync.rs:
```diff
@@ -883,42 +931,92 @@ impl RemoteTimelineClient {
+                self.launch_queued_tasks(upload_queue);
+            }
+            Outcome::ShutdownRequested => {
+                // Whichever *tasks* for this RemoteTimelineClient grabs the mutex first will transition the queue
+                // into shut down state, thereby dropping all off the queued *ops* which haven't become *tasks* yet.
+                // The other *tasks* will come here and observe an already shut down queue and hence simply wrap up their business.
```
What exactly is the "wrap up their business" here? ISTM that they might well finish the upload successfully, breaking the earlier loop with Outcome::TaskSucceeded. Then they will call `upload_queue.initialized_mut()` above, and panic because the queue was already in ShutDown stat.

problame replied:
Right, need a similar `match` in the `TaskSucceeded` case

problame replied:
Will address this tomorrow.

problame replied:
Done

hlinnaka commented:
> Right now, wait_completion() is only used by tests, but I
> suspect that we should be using it in wherever we
> delete layer files, e.g., GC and compaction, as explained
> in the storage_sync.rs block comment section "Consistency".

Yes, indeed, GC and compaction should call it. I'm pretty sure we had those calls in place at some point, not sure when they were lost..

hlinnaka commented:
> > Right now, wait_completion() is only used by tests, but I
> > suspect that we should be using it in wherever we
> > delete layer files, e.g., GC and compaction, as explained
> > in the storage_sync.rs block comment section "Consistency".
> 
> Yes, indeed, GC and compaction should call it. I'm pretty sure we had those calls in place at some point, not sure when they were lost..

This is a pre-existing bug, BTW. Even on 'main', there are no calls like `wait_completion` in timeline.rs. But I thought I had fixed that in PR 2785.

problame commented:
@hlinnaka addressed all code review comments.

I see you posted an alternate approach in https://github.com/neondatabase/neon/pull/2921 
Taking a look now.

problame commented:
Abandoned in favor of https://github.com/neondatabase/neon/pull/2921
